[DependencyInjection] Support autowiring for EventDispatcher/EventDispatcherInterface

chalasr opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

As it is a very common dependency. Currently it gives:

> [Symfony\Component\DependencyInjection\Exception\RuntimeException]
> Unable to autowire argument of type "Symfony\Component\EventDispatcher\EventDispatcherInterface" for the service "dummy". Multiple services exist for this interface (debug.event_dispatcher, debug.event_dispatcher.parent).

After this, the `TraceableEventDispatcher` will be injected in dev and the `ContainerAwareEventDispatcher` in prod, as when injecting `@event_dispatcher` explicitly.

ping @weaverryan

IMHO this could be treated as a an enhancement for the autowiring feature and be part of 3.2.

chalasr committed:
Support autowiring for EventDispatcher/EventDispatcherInterface

CS Fixes

chalasr commented:
The `TraceableEventDispatcher` (`debug.event_dispatcher` service)  must be the injected one in dev environment as it decorates it. However the current implementation gives always the decorated service (`event_dispatcher`), not the decorator.

Status: needs work

ogizanagi commented:
The `TraceableEventDispatcher` is not registered in the container by using the `decorates` feature of the DIC, but instead by doing this weird thing in the [`FrameworkExtension`](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L452-L456).

If we want to make this PR work properly with the event dispatcher in debug mode, we should transfer the `autowiring_type` to the `debug.event_dispatcher` service definition...
Or maybe we should remove the code from the `FrameworkExtension` and use the `decorates` feature instead ? It'll allow to handle correctly the autowiring type with decorated event dispatcher instance more straightforwardly. But I don't know what would be the impacts and if it can cause any BC break...?

chalasr commented:
@ogizanagi As you suggested, I made `@debug.event_dispatcher` decorates `@event_dispatcher` and removed the definition replacement logic from the `FrameworkExtension`. 
I added some tests.

The `debug.event_dispatcher.parent` service doesn't exist anymore, it is now `debug.event_dispatcher.inner`. AFAIK it should not be a BC break as the service was just created internally to mimic what `decorates` does

However we still need to manually cleanup the`event_dispatcher` autowiring types [from the extension](https://github.com/symfony/symfony/pull/20260/files#diff-0e793081ceb720201745c982a568903fR453), otherwise it is always a `ContainerAwareEventDispatcher` (the decorated) which is injected. Maybe the `AutowirePass` has aliases in its map (the decorated `@event_dispatcher`)?

Anyway, that's really not a big deal (one method call on the definition).

chalasr commented:
Status: needs review

chalasr commented:
Failing build on travis unrelated (yaml)

ogizanagi commented:
> However we still need to manually cleanup the event_dispatcher autowiring types from the extension, otherwise it is always a ContainerAwareEventDispatcher (the decorated) which is injected. Maybe the AutowirePass has aliases in its map (the decorated @event_dispatcher)?
>
> Anyway, that's really not a big deal (one method call on the definition).

It's kind of a big deal. 😄 I thought 71d502a was supposed to handle this case 😕

chalasr commented:
Needs #20267 for removing [this method call](https://github.com/symfony/symfony/pull/20260/files#diff-0e793081ceb720201745c982a568903fR453).

chalasr commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/debug.xml:
```diff
@@ -9,11 +9,13 @@
             <argument type="service" id="debug.stopwatch" />
             <argument type="service" id="logger" on-invalid="null" />
+            <autowiring-type>Symfony\Component\EventDispatcher\EventDispatcherInterface</autowiring-type>
+            <autowiring-type>Symfony\Component\EventDispatcher\EventDispatcher</autowiring-type>
```
This will not be needed with #20267 too

chalasr commented:
Needs branch 2.8 to be merged in master

chalasr commented:
I made the changes after #20267, this is ready.

chalasr commented:
Tests have been fixed.

xabbuh commented:
👍 

Status: Reviewed

javiereguiluz commented:
@dunglas could you please review/approve this proposal? Do you think it's ready for 3.2? Thanks!

dunglas commented:
:+1:

For me it's ok to merge this PR in 3.2.

chalasr commented:
apparently it will finally be for 3.3 as many others, it would be nice to have an explanation.

dunglas commented:
@chalasr sorry about that, the core team is very busy, reviewing can take time. Then @fabpot is the only one to do the tedious work of merging most PRs and preparing stable releases. It can, again, take some time.

fabpot commented:
End of development happened more than a month and a half ago. Since then, we should not have merged new features or significant changes. We did some for features that were discussion since a very long time, but we had priorities. RC1 is just around the corner, so merging anything substantial at this point is not possible anymore. Sorry about that, but we need to make choices. That's our release process. Date is fixed, features are not.

chalasr commented:
And I can perfectly understand, thank you for the hint.

chalasr commented:
I believe this can be merged now that 3.2 is released.

fabpot commented:
Thank you @chalasr.
